nomad/structs/group.go

```diff
@@ -0,0 +1,122 @@
+)
+
+var (
+	// Disconnect strategy validation errors
+	errStopAndLost         = errors.New("Disconnect cannot be configured with both lost_after and stop_after")
+	errNegativeLostAfter   = errors.New("lost_after cannot be a negative duration")
+	errNegativeStopAfter   = errors.New("stop_after cannot be a negative duration")
+	errStopAfterNonService = errors.New("stop_after can only be used with service or batch job types")
+	errInvalidReconcile    = errors.New("reconcile option is invalid")
+)
+
+func NewDefaultDisconnectStrategy() *DisconnectStrategy {
+	return &DisconnectStrategy{
+		Replace:   pointer.Of(true),
+		Reconcile: ReconcileOptionBestScore,
+	}
+}
+
+// Disconnect strategy defines how both clients and server should behave in case of
+// disconnection between them.
+type DisconnectStrategy struct {
+	// Defines for how long the server will consider the unresponsive node as
+	// disconnected but alive instead of lost.
+	LostAfter time.Duration `mapstructure:"lost_after" hcl:"lost_after,optional"`
+
+	// Defines for how long a disconnected client will keep its allocations running.
+	// This option has a different behavior for nil, the default, and time.Duration(0),
+	// and needs to be intentionally set/unset.
+	StopAfterOnClient *time.Duration `mapstructure:"stop_after_client" hcl:"stop_after_client, optional"`
+
+	// A boolean field used to define if the allocations should be replaced while
+	// its  considered disconnected.
+	// This option has a different behavior for nil, the default, and false,
+	// and needs to be intentionally set/unset. It needs to be set to true
+	// for compatibility.
+	Replace *bool `mapstructure:"replace" hcl:"replace,optional"`
+
+	// Once the disconnected node starts reporting again, it will define which
+	// instances to keep: the original allocations, the replacement, the one
+	// running on the node with the best score as it is currently implemented,
+	// or the allocation that has been running continuously the longest.
+	Reconcile string `mapstructure:"reconcile" hcl:"reconcile,optional"`
+}
+
+func (ds *DisconnectStrategy) Validate(job *Job) error {
+	var mErr *multierror.Error
+
+	if ds.StopAfterOnClient != nil {
+		if *ds.StopAfterOnClient < 0 {
+			mErr = multierror.Append(mErr, errNegativeStopAfter)
+		}
+
+		if job.Type != JobTypeService && job.Type != JobTypeBatch {
+			mErr = multierror.Append(mErr, errStopAfterNonService)
+		}
+	}
+
+	if ds.LostAfter < 0 {
+		mErr = multierror.Append(mErr, errNegativeLostAfter)
+	}
+
+	if ds.StopAfterOnClient != nil && ds.LostAfter != 0 {
+		return multierror.Append(mErr, errStopAndLost)
+	}
+
+	if ds.Reconcile != "" &&
+		ds.Reconcile != ReconcileOptionBestScore &&
+		ds.Reconcile != ReconcileOptionLongestRunning &&
+		ds.Reconcile != ReconcileOptionKeepOriginal &&
+		ds.Reconcile != ReconcileOptionKeepReplacement {
+		return multierror.Append(mErr, fmt.Errorf("%w: %s", errInvalidReconcile, ds.Reconcile))
+	}
```

Comment:
Another way of doing this is that I kinda like because it's almost got the readability of pattern-matching is:

```suggestion
	switch ds.Reconcile {
		case "", ReconcileOptionBestScore, ReconcileOptionLongestRunning, ReconcileOptionKeepOriginal, ReconcileOptionKeepReplacement:
		default:
			return multierror.Append(mErr, fmt.Errorf("%w: %s", errInvalidReconcile, ds.Reconcile))
		}
	}
```

But up to you.

Reply:
I particularly hate the switches with one case, personal preference jejej